applications/ParticleMechanicsApplication/custom_conditions/particle_based_conditions/mpm_particle_lagrange_dirichlet_condition.cpp

```diff
@@ -0,0 +1,458 @@
+        }
+
+        noalias( rRightHandSideVector ) = ZeroVector( matrix_size ); //resetting RHS
+    }
+
+    // Get imposed displacement and normal vector
+    const array_1d<double, 3 > & xg_c = this->GetValue(MPC_COORD);
+    const array_1d<double, 3 > & imposed_displacement = this->GetValue (MPC_IMPOSED_DISPLACEMENT);
+
+    // Prepare variables
+    GeneralVariables Variables;
+    const double augmentation_factor = this->GetValue(SCALAR_LAGRANGE_MULTIPLIER);
+    bool stabilization = this->GetValue(STABILIZATION_LAGRANGE_MULTIPLIER);
+
+    // Calculating shape function
+    Variables.N = this->MPMShapeFunctionPointValues(Variables.N, xg_c);
+    Variables.CurrentDisp = this->CalculateCurrentDisp(Variables.CurrentDisp, rCurrentProcessInfo);
+
+    // Check contact: Check contact penetration: if <0 apply constraint, otherwise no
+    bool apply_constraints = true;
+    if (Is(CONTACT))
+    {
+        // NOTE: the unit_normal_vector is assumed always pointing outside the boundary
+        array_1d<double, 3 > & unit_normal_vector = this->GetValue(MPC_NORMAL);
+        ParticleMechanicsMathUtilities<double>::Normalize(unit_normal_vector);
+        array_1d<double, 3 > field_displacement = ZeroVector(3);
+        for ( unsigned int i = 0; i < number_of_nodes; i++ )
+        {
+            if (Variables.N[i] > std::numeric_limits<double>::epsilon() )
+            {
+                for ( unsigned int j = 0; j < dimension; j++)
+                {
+                    field_displacement[j] += Variables.N[i] * Variables.CurrentDisp(i,j);
+                }
+            }
+        }
+
+        const double penetration = MathUtils<double>::Dot((field_displacement - imposed_displacement), unit_normal_vector);
+
+        // If penetrates, apply constraint, otherwise no
+        if (penetration >= 0.0)
+        {
+            apply_constraints = false;
+        }
+    }
+
+    if (apply_constraints)
+    {
+        Matrix lagrange_matrix = ZeroMatrix(matrix_size, matrix_size);
+
+         // Loop over Lagrange Multipliers
+        for (unsigned int i = 0; i < number_of_nodes; i++)
+        {
+            // Loop over shape functions of displacements
+            for (unsigned int j = 0; j < number_of_nodes; j++)
+            {
+                const double NN = Variables.N[j] * Variables.N[i];
+                const unsigned int ibase = i * dimension + dimension * number_of_nodes;
+                const unsigned int jbase = j * dimension;
+
+                // Matrix in following shape:
+                // |aug*H   H^T|
+                // |H       0  |
+                // Lambda in X
+                for (unsigned int k = 0; k < dimension; k++)
+                {
+                    lagrange_matrix(jbase+k, jbase+k) = NN * augmentation_factor;
+                    lagrange_matrix(ibase+k, jbase+k) = NN;
+                    lagrange_matrix(jbase+k, ibase+k) = NN;
+                }
+
+
+                //Stabilization needed for some solvers
```

Comment:
This is not an stabilization, you are indirectly applying ALM method, but it is not consistent if there is not RHS counterpart